File: android/src/main/kotlin/com/loopit/minis/telemetry/Telemetry.kt
package com.loopit.minis.telemetry

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Native side of `loopit/minis/telemetry`.
 *
 * Off by default. Hosts call `enable({sink: "log" | "stream"})` to turn it
 * on. Other engines inside the plugin (camera, videdit, audio, etc.) call
 * [Telemetry.publish] to emit events; this object holds the routing.
 *
 * Field scrubbing matches the Dart side allow-list — only structural
 * fields pass through, never paths/URIs/raw bytes.
 */
class Telemetry(messenger: BinaryMessenger) {

  private val main = Handler(Looper.getMainLooper())
  private val method = MethodChannel(messenger, "loopit/minis/telemetry")
  private val events = EventChannel(messenger, "loopit/minis/telemetry/events")
  private var eventSink: EventChannel.EventSink? = null

  init {
    instance = this

    method.setMethodCallHandler { call: MethodCall, result: MethodChannel.Result ->
      try {
        when (call.method) {
          "enable" -> {
            val sink = call.argument<String>("sink") ?: "log"
            enabled = true
            this.sinkMode = if (sink == "stream") SinkMode.STREAM else SinkMode.LOG
            result.success(null)
          }
          "disable" -> {
            enabled = false
            result.success(null)
          }
          "emit" -> {
            val name = call.argument<String>("name") ?: ""
            @Suppress("UNCHECKED_CAST")
            val raw = (call.argument<Map<String, Any?>>("fields") ?: emptyMap())
            publish(name, raw)
            result.success(null)
          }
          else -> result.notImplemented()
        }
      } catch (t: Throwable) {
        // Telemetry must never crash the host.
        result.error("telemetry-error", t.message, null)
      }
    }

    events.setStreamHandler(object : EventChannel.StreamHandler {
      override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) {
        eventSink = sink
      }
      override fun onCancel(arguments: Any?) { eventSink = null }
    })
  }

  fun dispose() {
    method.setMethodCallHandler(null)
    events.setStreamHandler(null)
    eventSink = null
    if (instance === this) {
      instance = null
    }
  }

  // ----------------------------- Routing ----------------------------- //

  private enum class SinkMode { LOG, STREAM }

  @Volatile private var enabled: Boolean = false
  @Volatile private var sinkMode: SinkMode = SinkMode.LOG

  private fun deliver(name: String, fields: Map<String, Any?>) {
    if (!enabled) return
    when (sinkMode) {
      SinkMode.LOG -> android.util.Log.i("MinisTelemetry", "$name $fields")
      SinkMode.STREAM -> {
        val payload = mapOf(
          "name" to name,
          "ts" to System.currentTimeMillis(),
          "fields" to fields,
        )
        main.post {
          try {
            eventSink?.success(payload)
          } catch (_: Throwable) {}
        }
      }
    }
  }

  companion object {
    @Volatile private var instance: Telemetry? = null

    /** Engines inside the plugin call this to emit a telemetry event. */
    @JvmStatic
    fun publish(name: String, fields: Map<String, Any?> = emptyMap()) {
      val ref = instance ?: return
      ref.deliver(name, scrub(fields))
    }

    private val ALLOW = setOf(
      "op", "kind", "durationMs", "pct", "fps", "codec", "profile",
      "width", "height", "sampleRate", "channels", "bitrateKbps",
      "thermal", "battery", "state", "result", "errorCode",
      "taskId", "segments", "count", "reason",
    )

    private fun scrub(input: Map<String, Any?>): Map<String, Any?> {
      val out = LinkedHashMap<String, Any?>(input.size)
      for ((k, v) in input) {
        if (k !in ALLOW) continue
        if (v is String && v.length > 64) continue
        out[k] = v
      }
      return out
    }
  }
}

File: ios/Classes/Telemetry/Telemetry.swift
import Flutter
import Foundation
import os.log

/// Native side of `loopit/minis/telemetry`.
///
/// Off by default. Hosts call `enable({sink: "log" | "stream"})` to turn it
/// on. Other engines in the plugin (camera, videdit, audio, etc.) call
/// `MinisTelemetry.publish(_:fields:)` to emit events; this object handles
/// routing.
///
/// Field scrubbing matches the Dart allow-list — only structural fields
/// pass through, never paths/URIs/raw bytes.
@objc public class MinisTelemetry: NSObject, FlutterStreamHandler {

  private enum SinkMode { case log, stream }

  // MARK: - Singleton wiring

  @objc public static let shared = MinisTelemetry()

  private var method: FlutterMethodChannel?
  private var events: FlutterEventChannel?
  private var sink: FlutterEventSink?

  private var enabled = false
  private var mode: SinkMode = .log
  private let log = OSLog(subsystem: "com.loopit.minis", category: "telemetry")
  private let queue = DispatchQueue(label: "loopit.minis.telemetry")

  @objc public func attach(messenger: FlutterBinaryMessenger) {
    let method = FlutterMethodChannel(
      name: "loopit/minis/telemetry",
      binaryMessenger: messenger
    )
    method.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }

    let events = FlutterEventChannel(
      name: "loopit/minis/telemetry/events",
      binaryMessenger: messenger
    )
    events.setStreamHandler(self)

    self.method = method
    self.events = events
  }

  @objc public func detach() {
    queue.sync {
      enabled = false
      sink = nil
    }
    method?.setMethodCallHandler(nil)
    events?.setStreamHandler(nil)
    method = nil
    events = nil
  }

  // MARK: - MethodChannel

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "enable":
      let args = call.arguments as? [String: Any]
      let sinkName = (args?["sink"] as? String) ?? "log"
      queue.sync {
        enabled = true
        mode = (sinkName == "stream") ? .stream : .log
      }
      result(nil)

    case "disable":
      queue.sync { enabled = false }
      result(nil)

    case "emit":
      let args = call.arguments as? [String: Any]
      let name = (args?["name"] as? String) ?? ""
      let fields = (args?["fields"] as? [String: Any]) ?? [:]
      MinisTelemetry.publish(name, fields: fields)
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - StreamHandler

  public func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
    queue.sync { sink = eventSink }
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    queue.sync { sink = nil }
    return nil
  }

  // MARK: - Publish (called by engines)

  /// Engines inside the plugin call this to emit a telemetry event.
  @objc public static func publish(_ name: String, fields: [String: Any] = [:]) {
    shared.deliver(name: name, fields: scrub(fields))
  }

  private func deliver(name: String, fields: [String: Any]) {
    var enabledSnapshot = false
    var modeSnapshot: SinkMode = .log
    var sinkSnapshot: FlutterEventSink?

    queue.sync {
      enabledSnapshot = enabled
      modeSnapshot = mode
      sinkSnapshot = sink
    }

    guard enabledSnapshot else { return }

    switch modeSnapshot {
    case .log:
      os_log("%{public}@ %{public}@", log: log, type: .info, name, "\(fields)")
    case .stream:
      let payload: [String: Any] = [
        "name": name,
        "ts": Int(Date().timeIntervalSince1970 * 1000.0),
        "fields": fields,
      ]
      DispatchQueue.main.async {
        sinkSnapshot?(payload)
      }
    }
  }

  // MARK: - Scrub

  private static let allow: Set<String> = [
    "op", "kind", "durationMs", "pct", "fps", "codec", "profile",
    "width", "height", "sampleRate", "channels", "bitrateKbps",
    "thermal", "battery", "state", "result", "errorCode",
    "taskId", "segments", "count", "reason",
  ]

  private static func scrub(_ input: [String: Any]) -> [String: Any] {
    var out: [String: Any] = [:]
    for (k, v) in input {
      guard allow.contains(k) else { continue }
      if let s = v as? String, s.count > 64 { continue }
      out[k] = v
    }
    return out
  }
}
